src/javaServerStarter.ts

```diff
@@ -1,24 +1,48 @@
-import { workspace, ExtensionContext } from 'vscode'
+import { workspace, ExtensionContext, window } from 'vscode'
 import { Executable, ExecutableOptions } from 'vscode-languageclient';
 import { RequirementsData } from './requirements';
 import * as os from 'os';
 import * as path from 'path';
 import { xmlServerVmargs, getJavaagentFlag, getKey, IS_WORKSPACE_VMARGS_XML_ALLOWED, getXMLConfiguration } from './settings';
+import * as http from 'http';
+import * as fs from 'fs';
+import { createHash, Hash } from 'crypto';
 const glob = require('glob');
 
 declare var v8debug;
 
 const DEBUG = (typeof v8debug === 'object') || startedInDebugMode();
 
-export function prepareExecutable(requirements: RequirementsData, xmlJavaExtensions: string[], context: ExtensionContext): Executable {
+export async function prepareExecutable(requirements: RequirementsData, xmlJavaExtensions: string[], context: ExtensionContext): Promise<Executable> {
   let executable: Executable = Object.create(null);
   let options: ExecutableOptions = Object.create(null);
   options.env = process.env;
   options.stdio = 'pipe';
   executable.options = options;
-  executable.command = path.resolve(requirements.java_home + '/bin/java');
-  executable.args = prepareParams(requirements, xmlJavaExtensions, context);
-  return executable;
+  const useBinary = getXMLConfiguration().get("server.binary.enabled", false);
+  const javaExecutable: () => Executable = (): Executable => {
+    executable.command = path.resolve(requirements.java_home + '/bin/java');
+    executable.args = prepareParams(requirements, xmlJavaExtensions, context);
+    return executable;
+  }
+  if (useBinary) {
+    //TODO handle case when other extensions contributing jars expect to force JVM mode
+    // TODO: handle arguments to the binary
+    return getServerBinary().then((serverBinary: string): Executable => {
+      const hash: Hash = createHash('sha256');
+      hash.update(fs.readFileSync(path.resolve(serverBinary)));
+      const hashDigest = hash.digest('hex').toLowerCase();
+      const expectedHashPath = path.resolve(__dirname, '../server', `lemminx-${os.platform}-hash`);
+      const expectedHash = fs.readFileSync(expectedHashPath).toString('utf-8').split(' ')[0];
+      if (hashDigest === expectedHash) {
+        executable.command = serverBinary;
+        return executable;
+      }
+      window.showErrorMessage(`The server binary ${serverBinary} has been modified! Falling back to the Java server...`);
```

Comment:
better ask the user if they want to proceed, in case of integrity check failure (at their own risks). Easier to test the binary that way without needing to regenerate the hash every time